### PR TITLE
inference: create a separate type for doing optimizations

### PR DESCRIPTION
**Pull request by vtjnash**
Most fields of InferenceState aren't valid during optimization,
so the goal is to reflect that in the structure of the types.

Similarly, doing optimization operations during inference would be
invalid, so this helps distinguish those cases as well.

Finally, this makes is possible for non-InferenceState-initiated IR passes
(e.g. external to typeinf) to make uses of these passes.

(Also, since it may not be eminently clear, note that the other goal of starting this is to eventually separate anything that takes an InferenceState into one file, and anything that takes OptimizationState into a second file)

### CODE REVIEW

File: base/inference.jl [hunk 1]
```diff
@@ -2816,11 +2881,24 @@ coverage_enabled() = (JLOptions().code_coverage != 0)
...
     sv.min_valid = max(sv.min_valid, min_valid)
     sv.max_valid = min(sv.max_valid, max_valid)
-    @assert !isa(sv.linfo.def, Method) || !sv.cached || sv.min_valid <= sv.params.world <= sv.max_valid "invalid age range update"
+    @assert(!isa(sv.linfo.def, Method) ||
+            !sv.cached ||
+            sv.min_valid <= sv.params.world <= sv.max_valid,
+            "invalid age range update")
+    nothing
+end
+function update_valid_age!(min_valid::UInt, max_valid::UInt, sv::OptimizationState)
```
Comment (tkelman) [inline base/inference.jl]: why isn't the mutated state input the first argument to this function?

Comment (vtjnash) [inline base/inference.jl]: consistency

Comment (tkelman) [inline base/inference.jl]: with what? that's an inconsistency with Julia API recommendations everywhere else

Comment (vtjnash) [inline base/inference.jl]: the rest of the file

Comment (tkelman) [inline base/inference.jl]: then

> why isn't the mutated state input the first argument to this function

in the rest of the file?

File: base/inference.jl [hunk 2]
```diff
@@ -2742,19 +2806,31 @@ end
 
 #### helper functions for typeinf initialization and looping ####
 
+# scan body for the value of value of the largest referenced label
```
Comment (tkelman) [inline base/inference.jl]: "value of value" intended?

File: base/inference.jl (hunk 1 above)
Comment (vtjnash) [inline base/inference.jl]: probably convention or something

Comment (JeffBezanson) [inline base/inference.jl]: These conventions are not really important for internal functions.

File: base/inference.jl [hunk 3]
```diff
@@ -308,6 +306,71 @@ function get_staged(li::MethodInstance)
...
+    src::CodeInfo
+    mod::Module
+    nargs::Int
+    next_label::Int # index of the current highest label for this function
+    min_valid::UInt
+    max_valid::UInt
+    params::InferenceParams
+    function OptimizationState(frame::InferenceState)
+        s_edges = frame.stmt_edges[1]
+        if s_edges=== ()
```
Comment (JeffBezanson) [inline base/inference.jl]: operator spacing

File: base/inference.jl [hunk 4]
```diff
@@ -308,6 +306,71 @@ function get_staged(li::MethodInstance)
...
+            s_edges = []
+            frame.stmt_edges[1] = s_edges
```
Comment (JeffBezanson) [inline base/inference.jl]: Is there any way around mutating `frame`? That's a pretty strange thing for a constructor to do.

File: base/inference.jl (hunk 1 above)
Comment (JeffBezanson) [inline base/inference.jl]: Re: the confused emojis: yes it's nice to follow the conventions, but we have more important things to spend time on than reviewing the design of every private function that nobody will ever see. Priorities, priorities, priorities.

File: base/inference.jl [hunk 5]
```diff
@@ -3365,40 +3460,44 @@ function optimize(me::InferenceState)
     type_annotate!(me)
 
     # run optimization passes on fulltree
-    force_noinline = false
+    force_noinline = true
```
Comment (JeffBezanson) [inline base/inference.jl]: Why did this change?

File: base/inference.jl (hunk 1 above)
Comment (tkelman) [inline base/inference.jl]: most critical thing to fix right now, no, but it is actively confusing for API's to be backwards for anyone who's debugging around this file

File: base/inference.jl (hunk 4 above)
Comment (vtjnash) [inline base/inference.jl]: I could make it part of the initialization for frame, but it's just more sensible to handle the lazy initialization here.

File: base/inference.jl (hunk 5 above)
Comment (vtjnash) [inline base/inference.jl]: We've been miscomputing this; but fortunately, since it only affects the cached=no case, they're haven't been any consumers of the data.

Comment (JeffBezanson) [inline base/inference.jl]: It doesn't make sense to me that `force_noinline` would default to true. It's supposed to be true only for functions declared `@noinline`. In fact with this change I think it's now always true? Am I missing something?

Comment (StefanKarpinski) [inline base/inference.jl]: Regardless of what the correct default is here, this **should not be included in a large, unrelated pull request**. If this is wrong, it should be changed in a separate individual PR with a commit message that explains what was wrong and why the change is correct.

Comment (vtjnash) [inline base/inference.jl]: It's a bit of an abuse of the variable name (phrasing it in the positive would be more accurate), since it should be dependent on optimize=true. We use this later to decide whether to compute the inlining optimization for the function. While instead that operation should be separated out into optimization, like the rest of this PR is working towards enforcing.

Comment (StefanKarpinski) [inline base/inference.jl]: If you're changing the meaning of this to the complete opposite, can you please change the name of the variable?
